servers/xr/xr_hand_tracker.h

```diff
@@ -0,0 +1,137 @@
+
+#include "core/object/ref_counted.h"
+
+class XRHandTracker : public RefCounted {
+	GDCLASS(XRHandTracker, RefCounted);
+	_THREAD_SAFE_CLASS_
+
+public:
+	enum Hand {
+		HAND_LEFT,
+		HAND_RIGHT,
+		HAND_MAX,
+	};
+
+	enum HandTrackingSource {
+		HAND_TRACKING_SOURCE_UNKNOWN,
+		HAND_TRACKING_SOURCE_UNOBSTRUCTED,
+		HAND_TRACKING_SOURCE_CONTROLLER,
+		HAND_TRACKING_SOURCE_MAX
+	};
+
+	enum HandJoint {
+		HAND_JOINT_PALM,
+		HAND_JOINT_WRIST,
+		HAND_JOINT_THUMB_METACARPAL,
+		HAND_JOINT_THUMB_PHALANX_PROXIMAL,
+		HAND_JOINT_THUMB_PHALANX_DISTAL,
+		HAND_JOINT_THUMB_TIP,
+		HAND_JOINT_INDEX_FINGER_METACARPAL,
+		HAND_JOINT_INDEX_FINGER_PHALANX_PROXIMAL,
+		HAND_JOINT_INDEX_FINGER_PHALANX_INTERMEDIATE,
+		HAND_JOINT_INDEX_FINGER_PHALANX_DISTAL,
+		HAND_JOINT_INDEX_FINGER_TIP,
+		HAND_JOINT_MIDDLE_FINGER_METACARPAL,
+		HAND_JOINT_MIDDLE_FINGER_PHALANX_PROXIMAL,
+		HAND_JOINT_MIDDLE_FINGER_PHALANX_INTERMEDIATE,
+		HAND_JOINT_MIDDLE_FINGER_PHALANX_DISTAL,
+		HAND_JOINT_MIDDLE_FINGER_TIP,
+		HAND_JOINT_RING_FINGER_METACARPAL,
+		HAND_JOINT_RING_FINGER_PHALANX_PROXIMAL,
+		HAND_JOINT_RING_FINGER_PHALANX_INTERMEDIATE,
+		HAND_JOINT_RING_FINGER_PHALANX_DISTAL,
+		HAND_JOINT_RING_FINGER_TIP,
+		HAND_JOINT_PINKY_FINGER_METACARPAL,
+		HAND_JOINT_PINKY_FINGER_PHALANX_PROXIMAL,
+		HAND_JOINT_PINKY_FINGER_PHALANX_INTERMEDIATE,
+		HAND_JOINT_PINKY_FINGER_PHALANX_DISTAL,
+		HAND_JOINT_PINKY_FINGER_TIP,
+		HAND_JOINT_MAX,
+	};
+
+	enum HandJointFlags {
+		HAND_JOINT_FLAG_ORIENTATION_VALID = 1,
+		HAND_JOINT_FLAG_ORIENTATION_TRACKED = 2,
+		HAND_JOINT_FLAG_POSITION_VALID = 4,
+		HAND_JOINT_FLAG_POSITION_TRACKED = 8,
+		HAND_JOINT_FLAG_LINEAR_VELOCITY_VALID = 16,
+		HAND_JOINT_FLAG_ANGULAR_VELOCITY_VALID = 32,
+	};
+
+	void set_hand(Hand p_hand);
+	Hand get_hand() const;
+
+	void set_has_tracking_data(bool p_has_tracking_data);
+	bool get_has_tracking_data() const;
+
+	void set_hand_tracking_source(HandTrackingSource p_source);
+	HandTrackingSource get_hand_tracking_source() const;
+
+	void set_hand_joint_flags(HandJoint p_joint, BitField<HandJointFlags> p_flags);
+	BitField<HandJointFlags> get_hand_joint_flags(HandJoint p_joint) const;
+
+	void set_hand_joint_transform(HandJoint p_joint, const Transform3D &p_transform);
+	Transform3D get_hand_joint_transform(HandJoint p_joint) const;
+
+	void set_hand_joint_radius(HandJoint p_joint, float p_radius);
+	float get_hand_joint_radius(HandJoint p_joint) const;
+
+	void set_hand_joint_linear_velocity(HandJoint p_joint, const Vector3 &p_velocity);
+	Vector3 get_hand_joint_linear_velocity(HandJoint p_joint) const;
+
+	void set_hand_joint_angular_velocity(HandJoint p_joint, const Vector3 &p_velocity);
+	Vector3 get_hand_joint_angular_velocity(HandJoint p_joint) const;
+
+protected:
+	static void _bind_methods();
+
+private:
+	Hand hand = HAND_LEFT;
+	bool has_tracking_data = false;
+	HandTrackingSource hand_tracking_source = HAND_TRACKING_SOURCE_UNKNOWN;
+
+	BitField<HandJointFlags> hand_joint_flags[HAND_JOINT_MAX];
```

Comment:
Would it make sense to zero-initialize some of the arrays using `= { };` to prevent the user from possibly accessing uninitialized data if tracking isn't available?

Reply:
This should already be covered!

`BitField`, `Transform3D` and `Vector3` all have default constructors that will initialize their data. The only one that needs `= { }` is the `hand_joint_radii` which already has it.

I just double checked this with a little GDScript, and everything looked good:

```
	var tracker = XRHandTracker.new()
	for i in range(XRHandTracker.HAND_JOINT_MAX):
		print ("Tracker radius ", i, " : ", tracker.get_hand_joint_radius(i))
		print ("Tracker transform ", i, " : ", tracker.get_hand_joint_transform(i))
		print ("Tracker flags ", i, " : ", tracker.get_hand_joint_flags(i))
		print ("Tracker linear velocity ", i, " : ", tracker.get_hand_joint_linear_velocity(i))
		print ("Tracker angular velocity ", i, " : ", tracker.get_hand_joint_angular_velocity(i))
```